File: 백준/Gold/18808. 스티커 붙이기/스티커 붙이기.cc
#include <bits/stdc++.h>
using namespace std;



int n, m, k;
int board[41][41];
int sticker[11][11]; 
int r, c;

bool isavail(int x, int y) {
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            if (sticker[i][j] && board[x+i][y+j]) {
                //스티커 한 칸이 들어갈 수 없으면
                return false;
            }
        }
    }
    return true;
}

void stick(int x, int y) {
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            if (sticker[i][j]) {
                board[x + i][y + j] = 1;
            }
        }
    }
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> m >> k;

    for (int t = 0; t < k; t++) {

        cin >> r >> c;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                cin >> sticker[i][j];
            }
        }

        int count = 0;
        while (count < 5) {
            bool flag = false;
            for (int i = 0; i <= n - r; i++) {
                for (int j = 0; j <= m - c; j++) {
                    if (isavail(i, j)) {
                        stick(i, j);
                        flag = true;
                        break;
                    }
                }
                if (flag) break;
            }
            if (flag) break;

            if (!flag) { //회전
                count++;
                int tmp[11][11];
                for (int i = 0; i < r; i++) {
                    for (int j = 0; j < c; j++) {
                        tmp[j][r - i - 1] = sticker[i][j];
                    }
                }

                for (int i = 0; i < r; i++)
                    fill(sticker[i], sticker[i] + c, 0);
                swap(r, c);

                for (int i = 0; i < r; i++) {
                    for (int j = 0; j < c; j++) {
                        sticker[i][j] = tmp[i][j];
                    }
                }
            }
        }
    }
    int ans = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (board[i][j] == 1) ans++;
        }

    }
    cout << ans;
}

File: 백준/Gold/18808. 스티커 붙이기/스티커 붙이기.py
import math
import sys
import queue
import copy
from itertools import combinations

n, m, k = list(map(int, sys.stdin.readline().strip().split()))
notebook = [[0 for _ in range(41)] for _ in range(41)]

def isAvailable(x, y, sticker, r, c):
    for i in range(r):
        for j in range(c):
            if notebook[x+i][y+j] == 1 and sticker[i][j] == 1:
                return False
    return True

def rotate(sticker, r, c):
    tmp = [[0 for _ in range(11)] for _ in range(11)]
    for i in range(r):
        for j in range(c):
            tmp[j][r-i-1] = sticker[i][j]

    r, c = c, r
    for i in range(r):
        for j in range(c):
            sticker[i][j] = tmp[i][j]
    

for _ in range(k):
    r, c = list(map(int, sys.stdin.readline().strip().split()))
    sticker = [[0 for _ in range(11)] for _ in range(11)]
    for i in range(r):
        tmp = list(map(int, sys.stdin.readline().strip().split()))
        for j in range(c):
            sticker[i][j] = tmp[j]


    flag = False
    count = 0
    while count < 4:
        for i in range(n-r+1):
            for j in range(m-c+1):
                if isAvailable(i, j, sticker, r, c):
                    for k in range(r):
                        for l in range(c):
                            if sticker[k][l]:
                                notebook[i+k][j+l] = 1
         
                    flag = True
                    break
            if flag:
                break
        if flag:
            break

        rotate(sticker, r, c)
        r, c = c, r

        count += 1
ans = 0
for i in range(n):
    for j in range(m):
        if notebook[i][j] == 1:
            ans+=1
print(ans)